### PR TITLE
Reworking EventListener v.3b

### PR DESCRIPTION
**Pull request by RedstoneFuture**
**Reworked (Refactoring, Improvement the checks, Continuity check) and tested event listener Ingame**

- PlayerArenaJoinEvent PL LL GL EL
- PlayerMoveEvent PL GL
- PlayerArenaLeaveEvent PL LL GL EL
- PlayerQuitEvent PL
- PlayerJoinEvent PL
- PlayerTeleportEvent PL
- PlayerMoveEvent PL, GL
- PrePlayerArenaJoinEvent (PL) = deleted

---

EL: EndListener
GL: GameListener
LL: LobbyListener
PL: PlayerListener

### CODE REVIEW

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java [hunk 1]
```diff
@@ -92,117 +84,125 @@ public void onPickup(PlayerPickupItemEvent event) {
...
+        Player player = event.getPlayer();
+
+        // old game handling:
+        if (gameFrom != null) registerPlayerArenaLeaveEvent(player, gameFrom);
+
+        // teleport after a delay between the arena leave and the next area join
+        new BukkitRunnable() {
+            public void run() {
+                // new game handling:
+                if (gameTo != null) {
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java]: Probably move this if statement up and check before the bukkit runnable via a guard statement

```
                if (gameTo == null) return;
```

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java [hunk 2]
```diff
@@ -92,117 +84,125 @@ public void onPickup(PlayerPickupItemEvent event) {
...
 
-        game.removePlayer(mwPlayer);
+        // old game handling:
+        if (gameFrom != null) registerPlayerArenaLeaveEvent(player, gameFrom);
 
-        MissileWars.getInstance().getSignRepository().getSigns(game).forEach(MWSign::update);
+        // new game handling:
+        if (gameTo != null) {
+            PlayerArenaJoinEvent joinEvent = registerPlayerArenaJoinEvent(player, gameTo);
+            if (!(joinEvent.isCancelled())) return;
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java]: Remove unnecessary brackets
```suggestion
            if (!joinEvent.isCancelled()) return;
```

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java [hunk 3]
```diff
@@ -92,117 +84,125 @@ public void onPickup(PlayerPickupItemEvent event) {
...
+        sendEventDebugMessage(player, game);
+        Logger.NORMAL.log(player.getName() + " left the MW game " + game.getLobby().getName());
 
-    @EventHandler
-    public void onPlayerArenaLeave(PlayerArenaLeaveEvent event) {
-        Logger.DEBUG.log("PlayerArenaLeaveEvent: " + event.getPlayer().getName());
+        return onLeaveGame;
     }
 
+    /**
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java:183]: Nice JavaDoc comment :+1:

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java [hunk 4]
```diff
@@ -92,117 +84,125 @@ public void onPickup(PlayerPickupItemEvent event) {
...
-            if (event.isCancelled()) player.sendMessage(MessageConfig.getMessage("not_enter_arena"));
-            return event.isCancelled();
-        }
-        if (from != null && to != from) {
-            PlayerArenaLeaveEvent event = new PlayerArenaLeaveEvent(player, from);
-            Bukkit.getPluginManager().callEvent(event);
-        }
-        return false;
-    }
+    private void sendEventDebugMessage(Player player, Game game) {
```
Comment (daniel-naegele) [inline missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java]: Maybe rename to logEventDebugMessage, but good idea to have more possibilities for debugging :+1: 

```suggestion
    private void logEventDebugMessage(Player player, Game game) {
```

**Review by daniel-naegele — CHANGES_REQUESTED** [review]

File: missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java (hunk 4 above)
Comment (RedstoneFuture) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java]: I don't know. You don't "log" the debug message. You send it.

Comment (daniel-naegele) [reply missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java]: Well we literally do

```java
        Logger.DEBUG.log("Location: " + player.getLocation());
        Logger.DEBUG.log("Current game amount: " + GameManager.getInstance().getGameAmount());
        Logger.DEBUG.log("Lobby: " + game.getLobby().getDisplayName());
        Logger.DEBUG.log("Arena: " + game.getArena().getDisplayName());
        Logger.DEBUG.log("Team 1: " + game.getTeam1());
        Logger.DEBUG.log("Team 2: " + game.getTeam2());
```

**Comment by Butzlabben** [comment]
And please rebase from master to resolve conflicts

**Pull request merged**